src/test/java/it/gov/pagopa/payment/service/qrcode/expired/QRCodeExpirationServiceImplIntegrationTest.java

```diff
@@ -0,0 +1,186 @@
+        AUTHORIZATION_EXPIRED_CHARGE_DATE = OFFSET_NOW.minusMinutes(authorizationMinutes+50L);
+        CANCEL_EXPIRED_CHARGE_DATE = OFFSET_NOW.minusMinutes(cancelMinutes+50L);
+
+        service = new QRCodeExpirationServiceImpl(authorizationExpiredService, cancelExpiredService);
+        storeData();
+    }
+
+    @AfterEach
+    void cleanUp() {
+        repository.deleteAllById(trxs.stream().map(TransactionInProgress::getId).toList());
+    }
+
+    @Test
+    void test() {
+        // waitFor expired trxs deleted from db
+        long[] count = {0};
+        TestUtils.waitFor(() -> (count[0]=repository.count()) <= extractIdsFromTrxsMap(validTrxs).size(),
+                () -> "Expected %d trxs, found %d".formatted(N/2, count[0]),
+                200,
+                1000);
+
+        // valid trxs still on db
+        List<TransactionInProgress> expected = validTrxs.values().stream().flatMap(List::stream).toList();
+        List<TransactionInProgress> result2 = repository.findAllById(extractIdsFromTrxsMap(validTrxs));
+        Assertions.assertEquals(sortAndCleanDates(expected), sortAndCleanDates(result2));
+
+        // verify call to rewardCalculator cancel for IDENTIFIED expired trxs
+        expiredTrxs.get(SyncTrxStatus.IDENTIFIED).forEach(t -> Mockito.verify(rewardCalculatorRestClientSpy).cancelTransaction(t.getId()));
+
+        // verify AUTHORIZED expired trxs to be notified in idpay-transaction queue
+        checkConfirmEvents();
+    }
+
+    private List<String> extractIdsFromTrxsMap(Map<SyncTrxStatus, List<TransactionInProgress>> trxsMap) {
+        return trxsMap.values().stream().flatMap(List::stream).map(TransactionInProgress::getId).toList();
+    }
+
+    private List<TransactionInProgress> sortAndCleanDates(Collection<TransactionInProgress> trxs) {
+        trxs.forEach(t -> t.setElaborationDateTime(null));
+        return trxs.stream().sorted(Comparator.comparing(TransactionInProgress::getId)).toList();
+    }
+
+    private void checkConfirmEvents() {
+        Set<TransactionInProgress> expectedEvents = Set.copyOf(expiredTrxs.get(SyncTrxStatus.AUTHORIZED).stream().map(this::trxInProgress2TrxOutcome).toList());
+
+        List<ConsumerRecord<String, String>> consumerRecords = kafkaTestUtilitiesService.consumeMessages(topicConfirmNotification, expectedEvents.size(), 15000);
+
+        Set<TransactionInProgress> eventsResult = consumerRecords.stream()
+                .map(r -> {
+                    TransactionOutcomeDTO out = TestUtils.jsonDeserializer(r.value(), TransactionOutcomeDTO.class);
+                    if (out.getStatus().equals(SyncTrxStatus.REWARDED)) {
+                        assertEquals(out.getMerchantId(), r.key());
+                    } else {
+                        assertEquals(out.getUserId(), r.key());
+                    }
+
+                    return out;
+                })
+                .collect(Collectors.toSet());
+
+        Assertions.assertEquals(
```

Comment:
before to clean and sort, we should assert that the elaborationDateTime has been updated..
collected the expected into a HashMap instead of Set, in order, inside the eventsResults stream, to get the previous instance by id and check elaborationDateTime > stored (maybe, it's better to store it as now().minusSeconds(1))